### ISSUE TITLE
LinearLifetimeChecker SIL verification found an ownership error in MandatoryInlining

### ISSUE THREAD

**Issue by tayloraswift** [apple/swift#61041]
```swift 
struct S<T> 
{
}
extension S<Int?> 
{
    mutating 
    func foo(x:inout [Int: Int])
    {
        for _:Int in 0 ... 1
        {
            if let y:Int? = 0 as Int??
            {
                if case 0? = y
                {
                    continue 
                }
            }
            else if case 0? = 0 as Int?
            {
                continue 
            }
            { _ in }(&x[0, default: 0])
        }
    }
}
```

```text 
$ ~/.swiftenv/versions/DEVELOPMENT-SNAPSHOT-2022-09-06-a/usr/bin/swiftc test5.swift 
error: compile command failed due to signal 6 (use -v to see invocation)
Begin Error in Function: '$s5test51SVAASiSgRszlE3foo1xySDySiADGz_tF'
Found outside of lifetime use!
Value: %35 = argument of bb1 : $Optional<@callee_guaranteed @substituted <τ_0_0> () -> @out τ_0_0 for <Optional<Int>>> // users: %137, %75, %82, %99, %100, %73
User:  br bb15(%34 : $Optional<@callee_guaranteed @substituted <τ_0_0> () -> @out τ_0_0 for <Optional<Int>>>, %101 : $Optional<@callee_guaranteed @substituted <τ_0_0> () -> @out τ_0_0 for <Optional<Int>>>) // id: %102
Block: bb14
End Error in Function: '$s5test51SVAASiSgRszlE3foo1xySDySiADGz_tF'
Found ownership error?!
triggering standard assertion failure routine
UNREACHABLE executed at /home/build-user/swift/lib/SIL/Verifier/LinearLifetimeCheckerPrivate.h:211!
Please submit a bug report (https://swift.org/contributing/#reporting-bugs) and include the project and the crash backtrace.
Stack dump:
0.      Program arguments: /home/ec2-user/.swiftenv/versions/DEVELOPMENT-SNAPSHOT-2022-09-06-a/usr/bin/swift-frontend -frontend -c -primary-file test5.swift -target x86_64-unknown-linux-gnu -disable-objc-interop -color-diagnostics -new-driver-path /home/ec2-user/.swiftenv/versions/DEVELOPMENT-SNAPSHOT-2022-09-06-a/usr/bin/swift-driver -empty-abi-descriptor -resource-dir /home/ec2-user/.swiftenv/versions/DEVELOPMENT-SNAPSHOT-2022-09-06-a/usr/lib/swift -module-name test5 -o /tmp/TemporaryDirectory.F6ckCD/test5-1.o
1.      Swift version 5.8-dev (LLVM a049441a0549936, Swift a973fe5375d3851)
2.      Compiling with the current language version
3.      While evaluating request ExecuteSILPipelineRequest(Run pipelines { Mandatory Diagnostic Passes + Enabling Optimization Passes } on SIL for test5)
4.      While running pass #78 SILModuleTransform "MandatoryInlining".
5.      While verifying SIL function "@$s5test51SVAASiSgRszlE3foo1xySDySiADGz_tF".
 for 'foo(x:)' (at test5.swift:7:5)
```

![bitmap](https://user-images.githubusercontent.com/2556986/189543603-c1488459-f265-4e78-857f-aeb168caf9a7.png)

i reproduced this on `DEVELOPMENT-SNAPSHOT-2022-08-18-a` and `DEVELOPMENT-SNAPSHOT-2022-09-06-a`.

**Comment by atrick** [comment]
rdar://100371230 (Issue #61041: LinearLifetimeChecker SIL verification found an ownership error in MandatoryInlining)

**Comment by kavon** [comment]
Looks like this got fixed, at least as of the release/5.10 branch. I'm adding test coverage here: https://github.com/apple/swift/pull/71109
